### PR TITLE
Handle circular dependency upon dependency removal

### PR DESCRIPTION
**Pull request by raejin**
**Summary**
This is a proposed fix for the issue I submitted https://github.com/facebook/metro/issues/500. For problem statement, please refer to the issue linked to understand the scope of the bug.

Currently Metro does not handle dependency removal correctly in the presence of certain circular dependency. Since we rely on Metro server to produce correct dependency graph in hopes to bundle only the needed modules. This is especially an issue when a dependency is removed from the entry point, however, due to it having circular dependency, the updated dependency graph will leave out all its dependencies in the graph.

I added the following two more test cases to ensure the correctness of the algorithm:

1. Remove `B` from `E`: `removes a dependency with transient cyclic dependency`
image

2. Remove `B` from `E`: `removes a cyclic dependency which is both inverse dependency and direct dependency`
image

3. Remove `B` from `E`: `removes a sub graph that has internal cyclic dependency`
image

Feel free to propose more tests to ensure the correctness of the algorithm. I'm aware that this may introduce more expensive graph updates for certain scenarios, but I believe that ensuring the correctness is far more important for dependency graph updates.

## Implementation
Previously, our implementation will stop removing circular dependency due to it having remaining `inverseDependencies`:

https://github.com/facebook/metro/blob/984aab84358aeef8de13ba86c33834a1c1bf6926/packages/metro/src/DeltaBundler/traverseDependencies.js#L328-L330

This can be illustrated by this example: 
image

When removing `B` from `E`, `B` will still have an inverse dependency `A` left. Henceforth, the rest of the graph remain untouched. My proposed solution is to have this async function `canSafelyRemoveFromParentModule` recursively checking all the inverse dependencies. In this example, we will look up inverse dependencies of `A` all the way to the end until there is no inverse dependency. We can only safely remove this dependency if and only if its end inverse dependency (in this case `A` will have `B` as its end inverse dependency) only has one path and the path is the same as the parent path. 

**Test plan**
- Updated 1 existing test, with 2 additional unit tests to ensure the correctness of the dependency removal logic handling with circular dependency.

```
yarn run jest packages/metro/src/DeltaBundler/__tests__/traverseDependencies-test.js
```

### CODE REVIEW

File: packages/metro/src/DeltaBundler/traverseDependencies.js [hunk 1]
```diff
@@ -309,12 +312,71 @@ async function addDependency<T>(
...
+  inverseDependencies: Set<string>,
+  parentModule: string,
+  graph: Graph<T>,
+): Promise<boolean> {
+  const result = getAllTopLevelInverseDependencies(
+    inverseDependencies,
+    parentModule,
+    graph,
+    '',
+    new Set(),
```
Comment (noahsug) [inline packages/metro/src/DeltaBundler/traverseDependencies.js]: Maybe make these args optional with default values

File: packages/metro/src/DeltaBundler/traverseDependencies.js [hunk 2]
```diff
@@ -309,12 +312,71 @@ async function addDependency<T>(
...
+      });
+      return acc;
+    }, new Set());
+}
+
+/**
+ * Given `inverseDependencies`, tracing back inverse dependencies to
+ * see if it only leads back to `parentModule`.
+ */
+async function canSafelyRemoveFromParentModule<T>(
```
Comment (noahsug) [inline packages/metro/src/DeltaBundler/traverseDependencies.js]: I think this function is synchronous? In which case we should remove `async`

File: packages/metro/src/DeltaBundler/traverseDependencies.js [hunk 3]
```diff
@@ -309,12 +312,71 @@ async function addDependency<T>(
...
+  visited.add(currModule);
+  if (!inverseDependencies.size || currModule === parentModule) {
+    return new Set([currModule]);
+  }
+
+  return Array.from(inverseDependencies)
+    .filter(inverseDep => graph.dependencies.has(inverseDep))
+    .reduce((acc, inverseDep) => {
+      const mod = graph.dependencies.get(inverseDep);
+      if (!mod) {
```
Comment (noahsug) [inline packages/metro/src/DeltaBundler/traverseDependencies.js]: won't this always be false due to the `filter` above?

File: packages/metro/src/DeltaBundler/traverseDependencies.js [hunk 4]
```diff
@@ -323,9 +385,17 @@ function removeDependency<T>(
...
   module.inverseDependencies.delete(parentModule.path);
 
-  // This module is still used by another modules, so we cannot remove it from
-  // the bundle.
-  if (module.inverseDependencies.size) {
+  // Even if there are modules still using parentModule, we want to ensure
+  // there isn't circular dependency. Thus, we check if it can be safely remove
+  // by tracing back the inverseDependencies.
+  if (
+    module.inverseDependencies.size &&
```
Comment (noahsug) [inline packages/metro/src/DeltaBundler/traverseDependencies.js]: Can do the `module.inverseDependencies.size` check in `canSafelyRemoveFromParentModule`

File: packages/metro/src/DeltaBundler/traverseDependencies.js (hunk 3 above)
Comment (cpojer) [reply packages/metro/src/DeltaBundler/traverseDependencies.js]: I assume this is to make Flow happy, which isn't able to keep track of that information.

**Comment by cpojer** [comment]
Thank you so much for this fix. I believe this looks good but do you mind answering @noahsug's questions? I am also curious if there is any measurable performance difference but overall I think it is probably fine – all this data is available in the graph already and no expensive I/O needs to be done for this. If it's less than ~100ms additional time spent for a graph of 10k modules with one change I think we can live with this.

File: packages/metro/src/DeltaBundler/traverseDependencies.js (hunk 2 above)
Comment (raejin) [reply packages/metro/src/DeltaBundler/traverseDependencies.js]: Ahh, nice catch :)

File: packages/metro/src/DeltaBundler/traverseDependencies.js (hunk 3 above)
Comment (raejin) [reply packages/metro/src/DeltaBundler/traverseDependencies.js]: Yes what @cpojer said

File: packages/metro/src/DeltaBundler/traverseDependencies.js (hunk 4 above)
Comment (raejin) [reply packages/metro/src/DeltaBundler/traverseDependencies.js]: I think this may make it clear on what `canSafelyRemoveFromParentModule` function intends to do, as way to signify that if `module.inverseDependencies.size` is non zero then we need additional check for all of the inverse dependencies.

**Comment by raejin** [comment]
@cpojer 
> Thank you so much for this fix. I believe this looks good but do you mind answering @noahsug's questions? I am also curious if there is any measurable performance difference but overall I think it is probably fine – all this data is available in the graph already and no expensive I/O needs to be done for this. If it's less than ~100ms additional time spent for a graph of 10k modules with one change I think we can live with this.

After running the initial solution with our entrypoint which amounts to 8730 modules, there was obvious performance bug with it (~10s for removing the entrypoint).

With more investigation, I implemented a memoized version which helps avoiding unnecessary DFS. With the updated approach, we're looking at 1~2 seconds by removing an entrypoint which has 8730 modules. I also added more tests around various edge cases that I came up with while debugging with our entrypoint.

File: packages/metro/src/DeltaBundler/traverseDependencies.js [hunk 5]
```diff
@@ -309,12 +318,107 @@ async function addDependency<T>(
...
+ */
+function canSafelyRemoveFromParentModule<T>(
+  inverseDependencies: Set<string>,
+  parentModule: string,
+  graph: Graph<T>,
+  canBeRemovedSafely: Set<string>,
+  delta: Delta,
+): boolean {
+  const visited = new Set();
+  const result = getAllTopLevelInverseDependencies(
```
Comment (noahsug) [inline packages/metro/src/DeltaBundler/traverseDependencies.js]: maybe rename `result` to `inverseDependencies`? Since result isn't actually the result the function returns

File: packages/metro/src/DeltaBundler/traverseDependencies.js [hunk 6]
```diff
@@ -309,12 +318,107 @@ async function addDependency<T>(
...
+  if (!result.size) {
+    /**
+     * This happens when parentModule and inverseDependencies has circular dependency,
+     * this will eventaully became an empty set due to `visited` Set being the base case
+     * for the recursive call.
+     */
+    return true;
+  }
+
+  const filterNotDeletedResult = Array.from(result).filter(
```
Comment (noahsug) [inline packages/metro/src/DeltaBundler/traverseDependencies.js]: could rename to something like `undeletedInverseDependencies`

File: packages/metro/src/DeltaBundler/traverseDependencies.js [hunk 7]
```diff
@@ -309,12 +318,107 @@ async function addDependency<T>(
...
+  /**
+   * We can only mark `visited` set of modules to be safely removable if
+   * 1. We do not have top level module to compare with parentModule,
+   *   this can happen when trying to see if we can safely remove from
+   *   a module that was deleted. This is why we filtered them out with `delta.deleted`
+   * 2. We have one top module and it is parentModule
+   *
+   */
+  const canSafelyRemove =
+    !filterNotDeletedResult.length ||
```
Comment (noahsug) [inline packages/metro/src/DeltaBundler/traverseDependencies.js]: I think we can remove this line

Comment (raejin) [reply packages/metro/src/DeltaBundler/traverseDependencies.js]: We want to set `canSafelyRemove` when
1. `filterNotDeletedResult` is empty
2. `filterNotDeletedResult` is not empty but only has one item.

I don't think we can remove this 🤔

Comment (noahsug) [reply packages/metro/src/DeltaBundler/traverseDependencies.js]: Woops I misread the logic. Definitely do not remove this line 😛

**Comment by cpojer** [comment]
@raejin 
> With more investigation, I implemented a memoized version which helps avoiding unnecessary DFS. With the updated approach, we're looking at 1~2 seconds by removing an entrypoint which has 8730 modules. I also added more tests around various edge cases that I came up with while debugging with our entrypoint.

Just wanted to clarify this is not the usual case for every change, only when removing a module that has 8k+ dependencies of its own, right? Removing a module here and there won't have a big performance impact, is that correct?

**Comment by raejin** [comment]
> @raejin
> 
> > With more investigation, I implemented a memoized version which helps avoiding unnecessary DFS. With the updated approach, we're looking at 1~2 seconds by removing an entrypoint which has 8730 modules. I also added more tests around various edge cases that I came up with while debugging with our entrypoint.
> 
> Just wanted to clarify this is not the usual case for every change, only when removing a module that has 8k+ dependencies of its own, right? Removing a module here and there won't have a big performance impact, is that correct?

Yes that is correct. Even with 8k module the latency is negligible in my opinion.
